diff --git CMakeLists.txt CMakeLists.txt
index 3417914..23ea630 100644
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -13,7 +13,7 @@ add_subdirectory(dune)
 add_subdirectory(m4)
 
 configure_file(${PROJECT_SOURCE_DIR}/dune/grid/spgrid/version.hh.in
-               ${PROJECT_BINARY_DIR}/dune/grid/spgrid/version.hh
+               ${PROJECT_SOURCE_DIR}/dune/grid/spgrid/version.hh
                @ONLY)
 
 finalize_dune_project(GENERATE_CONFIG_H_CMAKE)
